guest-tools/td_virsh_tool.sh

```diff
@@ -0,0 +1,224 @@
+            usage
+            exit 1
+            ;;
+        *) break ;;
+        esac
+        shift
+    done
+}
+
+set_input_paths() {
+    base_img_path=$(realpath ${BASE_IMG})
+    if [ $? -ne 0 ] || [ ! -f ${base_img_path} ]; then
+        echo "Set base TD qcow2 image path with TD_IMG"
+        exit 1
+    fi
+    xml_template_path=$(realpath ${XML_TEMPLATE})
+    if [ $? -ne 0 ] || [ ! -f ${xml_template_path} ]; then
+        echo "Set libvirt guest XML template path with XML_TEMPLATE"
+        exit 1
+    fi
+}
+
+set_domain() {
+    local domain_id=1
+    while $(virsh list --state-running --name |
+        grep -q "\-${domain_id}"); do
+        domain_id=$((domain_id + 1))
+        if [ ${domain_id} -gt ${MAX_DOMAINS} ]; then
+            echo "Error: exceeded max allowed guests."
+            domain="" # avoid destroying latest domain on_exit
+            exit 1
+        fi
+    done
+    domain=${DOMAIN_PREFIX}-${domain_id}
+}
+
+create_overlay_image() {
+    overlay_image_path=/tmp/overlay.${domain}.qcow2
+    qemu-img create \
+        -f qcow2 \
+        -F qcow2 \
+        -b ${base_img_path} \
+        ${overlay_image_path} >/dev/null
+}
+
+create_domain_xml() {
+    awk -v img_path=${base_img_path} \
+        -v domain=${domain} \
+        -v overlay_path=${overlay_image_path} '
+        {
+	gsub("BASE_IMG_PATH", img_path, $0); 
+	gsub("DOMAIN", domain, $0); 
+	gsub("OVERLAY_IMG_PATH", overlay_path, $0); 
+	print;
+        }
+	' ${xml_template_path} >/tmp/${domain}.xml
+}
+
+boot_vm() {
+    virsh define /tmp/${domain}.xml >/dev/null
+    virsh start ${domain} >/dev/null
+}
+
+echo_ssh_cmd() {
+    local host_port=$(
+        virsh \
+            qemu-monitor-command ${domain} \
+            --hmp info usernet |
+            awk '/HOST_FORWARD/ {print $4}'
+    )
+    local guest_cid=$(
+        virsh \
+            qemu-monitor-command ${domain} \
+            --hmp info qtree |
+            awk '/guest-cid/ {print $3}'
+    )
+    echo "Domain ${domain} running with vsock CID: ${guest_cid}," \
+        "ssh -p ${host_port} root@localhost"
+}
+
+destroy() {
+    local domain_to_destroy="${1}"
+
+    echo "Destroying domain ${domain_to_destroy}."
+
+    virsh shutdown ${domain_to_destroy} &>/dev/null
+    virsh shutdown --domain ${domain_to_destroy} &>/dev/null
+
+    echo "Waiting for VM to shutdown ..."
+    sleep 5
+
+    virsh destroy ${domain_to_destroy} &>/dev/null
+    virsh destroy --domain ${domain_to_destroy} &>/dev/null
+    virsh undefine ${domain_to_destroy} &>/dev/null
+
+    rm -f /tmp/{overlay.${domain_to_destroy}.qcow2,${domain_to_destroy}.xml}
+}
+
+clean_one() {
+    domains_all=$(virsh list --all --name)
+    if echo ${domains_all} | tr " " "\n" | grep -Fxq ${domain_to_clean}; then
+        destroy ${domain_to_clean}
+    else
+        echo "Unknown domain ${domain_to_clean}"
+        echo "Existing domains: $(echo ${domains_all} | tr -d "\n")"
+        exit 1
+    fi
+    exit 0
+}
```

Comment:
should we clean the qcow image ?

Reply:
I'm doing this [here](https://github.com/canonical/tdx/blob/42b851d24ebd9eb581efa640300b3e13e5c0fae1/guest-tools/td_virsh_tool.sh#L191).